node/src/components/chainspec_loader.rs

```diff
@@ -248,3 +270,282 @@ where
         }
     }
 }
+
+/// This struct can be parsed from a TOML-encoded chainspec file.  It means that as the
+/// chainspec format changes over versions, as long as we maintain the protocol config in this
+/// form at the top of the chainspec file, it can continue to be parsed as an `UpgradePoint`.
+#[derive(Deserialize)]
+struct UpgradePoint {
+    #[serde(rename = "protocol")]
+    pub(crate) protocol_config: ProtocolConfig,
+}
+
+impl UpgradePoint {
+    /// Parses a chainspec file at the given path as an `UpgradePoint`.
+    fn from_chainspec_path<P: AsRef<Path>>(path: P) -> Result<Self, Error> {
+        let bytes = utils::read_file(path.as_ref().join(&CHAINSPEC_NAME))
+            .map_err(Error::LoadUpgradePoint)?;
+        Ok(toml::from_slice(&bytes)?)
+    }
+}
+
+fn dir_name_from_version(version: &Version) -> PathBuf {
+    PathBuf::from(version.to_string().replace(".", "_"))
+}
+
+/// Iterates the given path, returning the subdir representing the greatest SemVer version.
+///
+/// Subdir names should be semvers with dots replaced with underscores.
+fn max_installed_version(dir: &Path) -> Result<Version, Error> {
+    let mut max_version = Version::new(0, 0, 0);
+    let mut read_version = false;
+    for entry in fs::read_dir(dir).map_err(|error| Error::ReadDir {
+        dir: dir.to_path_buf(),
+        error,
+    })? {
+        let path = match entry {
+            Ok(dir_entry) => dir_entry.path(),
+            Err(error) => {
+                debug!(dir=%dir.display(), %error, "bad entry while reading dir");
+                continue;
+            }
+        };
+
+        let subdir_name = match path.file_name() {
+            Some(name) => name.to_string_lossy().replace("_", "."),
+            None => continue,
+        };
+
+        let version = match Version::from_str(&subdir_name) {
+            Ok(version) => version,
+            Err(error) => {
+                trace!(%error, path=%path.display(), "failed to get a version");
```

Comment:
I think this might be promoted to `warn!`. Almost no one will run with `trace` log level and it will be helpful to know quickly why we weren't able to read the upgrades directory.

Reply:
I see there's a `warn!` later. Maybe that will be enough.